### PR TITLE
Handle UpdateExpressions in no-direct-mutation-state

### LINKED ISSUE #1386: no-direct-mutation-state rule doesn't handle UpdateExpressions
`this.state.foo++` does not trigger the `no-direct-mutation-state` rule.

`this.state.foo += 1` does - indicating this is likely just that `UpdateExpression`s aren't handled.

### PR DESCRIPTION
**Pull request by zpao**
This rule should also catch increment & decrement operators.

Fixes #1386

### CODE REVIEW

File: tests/lib/rules/no-direct-mutation-state.js [hunk 1]
```diff
@@ -70,6 +70,14 @@ ruleTester.run('no-direct-mutation-state', rule, {
       '}'
     ].join('\n')
   }, {
+    code: [
+      'class Hello extends React.Component {',
+      '  constructor() {',
+      '    this.state.foo = 1;',
```
Comment (ljharb) [inline tests/lib/rules/no-direct-mutation-state.js]: Could we also add a test that makes `++` invalid inside the constructor?

Comment (zpao) [inline tests/lib/rules/no-direct-mutation-state.js]: Oh shoot, I must have missed adding that line checking another thing… Sorry.

I had `this.state.foo++` right here as a valid case - we're in the constructor and while I think it's awkward, it's legal.

Comment (ljharb) [inline tests/lib/rules/no-direct-mutation-state.js]: I don't see why it should be legal; it's mutating `this.state` after it's assigned.

Comment (zpao) [inline tests/lib/rules/no-direct-mutation-state.js]: Assignment doesn't actually matter until we're out of the `constructor` method. That's when React takes over.

But that's also a tangential change to what I wanted to make here - I'm bringing `++` to parity with `=` (you have a test case immediately above this one showing that). If you want to change how you treat assignment as well, that's fine but I think you probably want to do that distinctly.

Comment (ljharb) [inline tests/lib/rules/no-direct-mutation-state.js]: Fair enough.

Comment (TheSavior) [inline tests/lib/rules/no-direct-mutation-state.js]: This has been my favorite find in our codebase where we were assigning to state.

```
this.setState({
  foo: this.state.foo++
});
```

Comment (jaaberg) [reply tests/lib/rules/no-direct-mutation-state.js]: @TheSavior (cc: @zpao )

I actually have a PR proposing a rule catching all references to `this.state` used in at `this.setState` in #1374. Does this rule cover that as well? If not, I would be really happy for feedback.

**Comment by zpao** [comment]
Do you need anything else from me here?

**Pull request merged**